Panic attempting to create records for Istio gateway host "*"

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

-->

**What happened**:

external-dns crashed when I added an Istio gateway for TCP traffic matching all hosts.

```
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
sigs.k8s.io/external-dns/registry.affixNameMapper.toTXTName(0x0, 0x0, 0x0, 0x0, 0x3a83caa, 0x1, 0xc000570f40, 0xc00086fb00)
	/sigs.k8s.io/external-dns/registry/txt.go:241 +0x127
sigs.k8s.io/external-dns/registry.(*TXTRegistry).ApplyChanges(0xc00066c150, 0x28a7b00, 0xc000569d70, 0xc000583200, 0x0, 0x0)
	/sigs.k8s.io/external-dns/registry/txt.go:141 +0x381
sigs.k8s.io/external-dns/controller.(*Controller).RunOnce(0xc00028d5f0, 0x28a7b00, 0xc000569d70, 0x3b11da0, 0x2252e01)
	/sigs.k8s.io/external-dns/controller/controller.go:152 +0x38a
sigs.k8s.io/external-dns/controller.(*Controller).Run(0xc00028d5f0, 0x28a7a40, 0xc0003e8240)
	/sigs.k8s.io/external-dns/controller/controller.go:189 +0x205
main.main()
	/sigs.k8s.io/external-dns/main.go:341 +0xd4f
```

The panic occurs here:

https://github.com/kubernetes-sigs/external-dns/blob/102228c18138cce9c4689df1271742e7ea9cecd0/registry/txt.go#L240-L241

The `endpointDNSName` in my case does not contain a `.` to split on (it is just `*`).

**What you expected to happen**:

external-dns does not try to create a record for that host.

**How to reproduce it (as minimally and precisely as possible)**:

Run external-dns with `--source=istio-gateway`, and add a gateway with a wildcard host (either `*` or `some-namespace/*`):

```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: my-ingress
spec:
  selector:
    app: my-ingress-gateway
  servers:
  - port:
      number: 80
      name: http2
      protocol: HTTP2
    hosts:
    - "*"
```




**Anything else we need to know?**:

The Istio [gateway docs](https://istio.io/latest/docs/reference/config/networking/gateway/#Server) describe the behaviour:

>Set the `dnsName` to `*` to select all `VirtualService` hosts from the specified namespace (e.g.,`prod/*`).

**Environment**:
- External-DNS version (use `external-dns --version`): v20200603-v0.7.2
- DNS provider: Route53
- Others: Istio 1.6.1

--- a/source/gateway.go
+++ b/source/gateway.go
@@ -300,7 +300,9 @@ func (sc *gatewaySource) hostNamesFromGateway(gateway networkingv1alpha3.Gateway
 				host = parts[1]
 			}
 
-			hostnames = append(hostnames, host)
+			if host != "*" {
+				hostnames = append(hostnames, host)
+			}
 		}
 	}
 	return hostnames, nil

--- a/source/gateway.go
+++ b/source/gateway.go
@@ -272,11 +272,6 @@ func (sc *gatewaySource) endpointsFromGateway(hostnames []string, gateway networ
 
 	providerSpecific, setIdentifier := getProviderSpecificAnnotations(annotations)
 
-	// Skip endpoints if we do not want entries from annotations
-	if !sc.ignoreHostnameAnnotation {
-		hostnames = append(hostnames, getHostnamesFromAnnotations(annotations)...)
-	}
-
 	for _, host := range hostnames {
 		endpoints = append(endpoints, endpointsForHostname(host, targets, ttl, providerSpecific, setIdentifier)...)
 	}
@@ -305,6 +300,11 @@ func (sc *gatewaySource) hostNamesFromGateway(gateway networkingv1alpha3.Gateway
 			}
 		}
 	}
+
+	if !sc.ignoreHostnameAnnotation {
+		hostnames = append(hostnames, getHostnamesFromAnnotations(gateway.Annotations)...)
+	}
+
 	return hostnames, nil
 }
 

--- a/source/gateway_test.go
+++ b/source/gateway_test.go
@@ -1094,6 +1094,43 @@ func testGatewayEndpoints(t *testing.T) {
 			},
 			expected: []*endpoint.Endpoint{},
 		},
+		{
+			title:           "gateways with wildcard host and hostname annotation",
+			targetNamespace: "",
+			lbServices: []fakeIngressGatewayService{
+				{
+					ips: []string{"1.2.3.4"},
+				},
+			},
+			configItems: []fakeGatewayConfig{
+				{
+					name:      "fake1",
+					namespace: "",
+					annotations: map[string]string{
+						hostnameAnnotationKey: "fake1.dns-through-hostname.com",
+					},
+					dnsnames: [][]string{{"*"}},
+				},
+				{
+					name:      "fake2",
+					namespace: "",
+					annotations: map[string]string{
+						hostnameAnnotationKey: "fake2.dns-through-hostname.com",
+					},
+					dnsnames: [][]string{{"some-namespace/*"}},
+				},
+			},
+			expected: []*endpoint.Endpoint{
+				{
+					DNSName: "fake1.dns-through-hostname.com",
+					Targets: endpoint.Targets{"1.2.3.4"},
+				},
+				{
+					DNSName: "fake2.dns-through-hostname.com",
+					Targets: endpoint.Targets{"1.2.3.4"},
+				},
+			},
+		},
 	} {
 		t.Run(ti.title, func(t *testing.T) {
 

--- a/source/gateway_test.go
+++ b/source/gateway_test.go
@@ -1072,6 +1072,28 @@ func testGatewayEndpoints(t *testing.T) {
 			},
 			ignoreHostnameAnnotation: true,
 		},
+		{
+			title:           "gateways with wildcard host",
+			targetNamespace: "",
+			lbServices: []fakeIngressGatewayService{
+				{
+					ips: []string{"1.2.3.4"},
+				},
+			},
+			configItems: []fakeGatewayConfig{
+				{
+					name:      "fake1",
+					namespace: "",
+					dnsnames:  [][]string{{"*"}},
+				},
+				{
+					name:      "fake2",
+					namespace: "",
+					dnsnames:  [][]string{{"some-namespace/*"}},
+				},
+			},
+			expected: []*endpoint.Endpoint{},
+		},
 	} {
 		t.Run(ti.title, func(t *testing.T) {
 

